pkg/cmd/pipeline/list.go

```diff
@@ -16,51 +16,169 @@ package pipeline
+				return printPipelineListObj(cmd.OutOrStdout(), p, f)
 			}
-			return printer.PrintObj(ps, cmd.OutOrStdout())
+
+			return printPipelineDetails(cmd.OutOrStdout(), p)
 		},
 	}
 	f.AddFlags(c)
 
 	return c
 }
 
-func list(p cli.Params) (*v1alpha1.PipelineList, error) {
+func printPipelineDetails(out io.Writer, p cli.Params) error {
+
+	cs, err := p.Clientset()
+	if err != nil {
+		return err
+	}
+
+	ps, prs, err := listPipelineDetails(cs, p.Namespace())
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to list pipelines from %s namespace\n", p.Namespace())
+		return err
+	}
+
+	if len(ps.Items) == 0 {
+		fmt.Fprintln(out, emptyMsg)
+		return nil
+	}
+
+	w := tabwriter.NewWriter(out, 0, 5, 3, ' ', tabwriter.TabIndent)
+	fmt.Fprintln(w, header)
+
+	for _, pipeline := range ps.Items {
+		pr, ok := prs[pipeline.Name]
+		if !ok {
+			fmt.Fprintf(w, body,
+				pipeline.Name,
+				formatted.Age(pipeline.CreationTimestamp, p.Time()),
+				blank,
+				blank,
+				blank,
+				blank,
+			)
+			continue
+		}
+
+		fmt.Fprintf(w, body,
+			pipeline.Name,
+			formatted.Age(pipeline.CreationTimestamp, p.Time()),
+			pr.Name,
+			formatted.Age(*pr.Status.StartTime, p.Time()),
+			formatted.Duration(pr.Status.StartTime, pr.Status.CompletionTime),
+			formatted.Condition(pr.Status.Conditions[0]),
+		)
+
+	}
+	return w.Flush()
+}
+
+func printPipelineListObj(w io.Writer, p cli.Params, f *cliopts.PrintFlags) error {
 	cs, err := p.Clientset()
+	if err != nil {
+		return err
+	}
+
+	ps, err := listAllPipelines(cs, p.Namespace())
+	if err != nil {
+		return err
+	}
+
+	return printer.PrintObject(w, ps, f)
+}
+
+func listAllPipelines(cs versioned.Interface, ns string) (*v1alpha1.PipelineList, error) {
+	c := cs.TektonV1alpha1().Pipelines(ns)
+
+	ps, err := c.List(metav1.ListOptions{})
 	if err != nil {
 		return nil, err
 	}
 
-	c := cs.TektonV1alpha1().Pipelines(p.Namespace())
-	return c.List(v1.ListOptions{})
+	// NOTE: this is required for -o json|yaml to work properly since
+	// tektoncd go client fails to set these; probably a bug
+	ps.GetObjectKind().SetGroupVersionKind(
+		schema.GroupVersionKind{
+			Version: "tekton.dev/v1alpha1",
+			Kind:    "PipelineList",
+		})
+	return ps, nil
+}
+
+type pipelineruns map[string]v1alpha1.PipelineRun
+
+func listPipelineDetails(cs versioned.Interface, ns string) (*v1alpha1.PipelineList, pipelineruns, error) {
+
+	ps, err := listAllPipelines(cs, ns)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(ps.Items) == 0 {
+		return ps, pipelineruns{}, nil
+	}
+
+	runs, err := cs.TektonV1alpha1().PipelineRuns(ns).List(metav1.ListOptions{})
+	if err != nil {
+		// TODO: may be just the pipeline details can be print
+		return nil, nil, err
+	}
+
+	latestRuns := pipelineruns{}
+
+	for _, run := range runs.Items {
+		pipelineName := run.Spec.PipelineRef.Name
+		r, ok := latestRuns[pipelineName]
+		if !ok {
+			latestRuns[pipelineName] = run
+			continue
+		}
+		if run.Status.StartTime.After(r.Status.StartTime.Time) {
```

Comment:
@vdemeester does this make more sense than `creationTimestamp` ?

Reply:
:thinking: I tend to prefer `creationTimestamp` — for me *last run* means the last one that got created (not the last one that started), but this might "change" in the future, we'll see :wink:)

Reply:
cool, will change to use creationTimestamp